src/python/WMCore/Storage/RucioFileCatalog.py

```diff
@@ -0,0 +1,231 @@
+#!/usr/bin/env python
+"""
+_RucioFileCatalog_
+
+Object to contain LFN to PFN mappings from a Rucio File Catalog
+and provide functionality to match LFNs against them
+
+Usage:
+
+given a storage description file, storage.json, execute readRFC. This will return
+a RucioFileCatalog instance that can be used to match LFNs to PFNs.
+"""
+
+from builtins import next, str, range
+from future.utils import viewitems
+
+from future import standard_library
+standard_library.install_aliases()
+
+import os
+import re
+
+import json
+
+from urllib.parse import urlsplit
+from xml.dom.minidom import Document
+
+from WMCore.Algorithms.ParseXMLFile import xmlFileToNode
+
+class RucioFileCatalog(dict):
+    """
+    _RucioFileCatalog_
+
+    Object that can map LFNs to PFNs based on contents of a Rucio
+    File Catalog
+    """
+
+    def __init__(self):
+        dict.__init__(self)
+        self['lfn-to-pfn'] = []
+        self['pfn-to-lfn'] = []
+        self.preferredProtocol = None  # attribute for preferred protocol
+
+    def addMapping(self, protocol, match, result,
+                   chain=None, mapping_type='lfn-to-pfn'):
+        """
+        _addMapping_
+
+        Add an lfn to pfn mapping to this instance
+
+        """
+        entry = {}
+        entry.setdefault("protocol", protocol)
+        entry.setdefault("path-match-expr", re.compile(match))
+        entry.setdefault("path-match", match)
+        entry.setdefault("result", result)
+        entry.setdefault("chain", chain)
+        self[mapping_type].append(entry)
+
+    def _doMatch(self, protocol, path, style, caller):
+        """
+        Generalised way of building up the mappings.
+        caller is the method from there this method was called, it's used
+        for resolving chained rules
+
+        Return None if no match
+
+        """
+        for mapping in self[style]:
+            if mapping['protocol'] != protocol:
+                continue
+            if mapping['path-match-expr'].match(path) or mapping["chain"] != None:
+                if mapping["chain"] != None:
+                    oldpath = path
+                    path = caller(mapping["chain"], path)
```

Comment:
I would keep these as is for now since it is the same code as main repo:
https://github.com/nhduongvn/WMCore/blob/747c70a8a6508765f460073ba9d288f9e3d97d77/src/python/WMCore/Storage/TrivialFileCatalog.py#L75